### PR TITLE
Add EligibilityService to BraveFederatedService

### PR DESCRIPTION
**Pull request by LorenzoMinto**
Add NetworkChange and PowerState Observers

Resolves https://github.com/brave/brave-browser/issues/19501

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_federated/eligibility_service.cc [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_federated/eligibility_service.h"
+
+#include <memory>
+#include <string>
```
Comment (moritzhaller) [inline components/brave_federated/eligibility_service.cc:9]: not needed

File: components/brave_federated/eligibility_service.cc [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_federated/eligibility_service.h"
+
+#include <memory>
```
Comment (moritzhaller) [inline components/brave_federated/eligibility_service.cc:8]: not needed

File: components/brave_federated/eligibility_service.cc [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#include <string>
+#include <utility>
```
Comment (moritzhaller) [inline components/brave_federated/eligibility_service.cc:10]: not needed

File: components/brave_federated/eligibility_service.cc [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "base/power_monitor/power_monitor.h"
+#include "base/time/time.h"
```
Comment (moritzhaller) [inline components/brave_federated/eligibility_service.cc:13]: not needed

File: components/brave_federated/eligibility_service.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_FEDERATED_ELIGIBILITY_SERVICE_H_
+#define BRAVE_COMPONENTS_BRAVE_FEDERATED_ELIGIBILITY_SERVICE_H_
+
+#include <memory>
```
Comment (moritzhaller) [inline components/brave_federated/eligibility_service.h:9]: not needed

File: components/brave_federated/eligibility_service.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#include <string>
```
Comment (moritzhaller) [inline components/brave_federated/eligibility_service.h:10]: not needed

File: components/brave_federated/eligibility_service.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "base/power_monitor/power_monitor.h"
+#include "base/time/time.h"
```
Comment (moritzhaller) [inline components/brave_federated/eligibility_service.h:13]: not needed

**Comment by iefremov** [comment]
@moritzhaller can we please put some details into the ticket?

File: components/brave_federated/eligibility_service.h [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+#ifndef BRAVE_COMPONENTS_BRAVE_FEDERATED_ELIGIBILITY_SERVICE_H_
+#define BRAVE_COMPONENTS_BRAVE_FEDERATED_ELIGIBILITY_SERVICE_H_
+
+#include "base/power_monitor/power_monitor.h"
+#include "net/base/network_change_notifier.h"
+
+namespace brave_federated {
+
+class EligibilityService
```
Comment (iefremov) [inline components/brave_federated/eligibility_service.h:14]: pls add comments for classes and important methods

File: components/brave_federated/eligibility_service.cc [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+#include "brave/components/brave_federated/eligibility_service.h"
+
+#include "base/power_monitor/power_monitor.h"
+#include "net/base/network_change_notifier.h"
+
+namespace brave_federated {
+
+EligibilityService::EligibilityService()
+    : is_on_battery_power_(
+          base::PowerMonitor::AddPowerStateObserverAndReturnOnBatteryState(
```
Comment (iefremov) [inline components/brave_federated/eligibility_service.cc:15]: I'd unsubscrive from both observers in the destructor for sanity

File: components/brave_federated/eligibility_service.h (hunk 8 above)
Comment (iefremov) [reply components/brave_federated/eligibility_service.h:14]: Even after reading the code it's not really clear - eligibility for what? :)

**Review by moritzhaller — APPROVED** [review]

**Review by iefremov — APPROVED** [review]

**Pull request merged**